File: boot-data-redis/src/main/java/com/example/demo/PostHandler.java
package com.example.demo;

import java.net.URI;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@Component
class PostHandler {

  private final PostRepository posts;

  public PostHandler(PostRepository posts) {
    this.posts = posts;
  }

  public Mono<ServerResponse> all(ServerRequest req) {
    return ServerResponse.ok().body(this.posts.findAll(), Post.class);
  }

  public Mono<ServerResponse> create(ServerRequest req) {
    return req.bodyToMono(Post.class)
        .flatMap(this.posts::save)
        .flatMap(p -> ServerResponse.created(URI.create("/posts/" + p.getId())).build());
  }

  public Mono<ServerResponse> get(ServerRequest req) {
    return this.posts.findById(req.pathVariable("id"))
        .flatMap(post -> ServerResponse.ok().body(Mono.just(post), Post.class))
        .switchIfEmpty(ServerResponse.notFound().build());
  }

  public Mono<ServerResponse> update(ServerRequest req) {

    return Mono
        .zip(
            (data) -> {
              Post p = (Post) data[0];
              Post p2 = (Post) data[1];
              p.setTitle(p2.getTitle());
              p.setContent(p2.getContent());
              return p;
            },
            this.posts.findById(req.pathVariable("id")),
            req.bodyToMono(Post.class)
        )
        .cast(Post.class)
        .flatMap(this.posts::save)
        .flatMap(post -> ServerResponse.noContent().build());

  }

  public Mono<ServerResponse> delete(ServerRequest req) {
    return ServerResponse.noContent().build(this.posts.deleteById(req.pathVariable("id")));
  }

}

File: kotlin-routes/src/main/kotlin/com/example/demo/PostHandler.kt
package com.example.demo

import org.springframework.http.MediaType
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.ServerResponse.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.net.URI
import java.time.Duration

class PostHandler(val posts: PostRepository) {


    fun all(req: ServerRequest): Mono<ServerResponse> {
        return ok().body(this.posts.findAll(), Post::class.java)
    }

    fun stream(req: ServerRequest): Mono<ServerResponse> {
        val postStream = Flux
                .zip(Flux.interval(Duration.ofSeconds(1)), this.posts.findAll())
                .map { it.t2 }
        return ok()
                .contentType(MediaType.TEXT_EVENT_STREAM)
                .body(
                        postStream,
                        Post::class.java
                )
    }

    fun create(req: ServerRequest): Mono<ServerResponse> {
        return req.bodyToMono(Post::class.java)
                .flatMap { post -> this.posts.save(post) }
                .flatMap { (id) -> created(URI.create("/posts/" + id)).build() }
    }

    fun get(req: ServerRequest): Mono<ServerResponse> {
        return this.posts.findById(req.pathVariable("id"))
                .flatMap { post -> ok().body(Mono.just(post), Post::class.java) }
                .switchIfEmpty(notFound().build())
    }

    fun update(req: ServerRequest): Mono<ServerResponse> {
        return this.posts.findById(req.pathVariable("id"))
                .zipWith(req.bodyToMono(Post::class.java))
                .map { it.t1.copy(title = it.t2.title, content = it.t2.content) }
                .flatMap { this.posts.save(it) }
                .flatMap { noContent().build() }
    }

    fun delete(req: ServerRequest): Mono<ServerResponse> {
        return this.posts.deleteById(req.pathVariable("id"))
                .flatMap { noContent().build() }
    }
}

